src/main/java/build/buildfarm/instance/AbstractServerInstance.java

```diff
@@ -445,9 +445,13 @@ protected ExecuteOperationMetadata expectExecuteOperationMetadata(
   }
 
   protected Action expectAction(Operation operation) {
```

Comment:
I don't see you doing anything upstream to handle the possible null value returned here, like you do for command in validateAction. Maybe it would be simpler for all expectX functions to either return the X or throw a Precondition failure. That way you only have to handle it in one place.